### PR TITLE
Added K8s custom resource definitions skeletons

### LINKED ISSUE #103: If we start with no listeners, warn and wait for new config
Currently, secretless exits when there are no listeners configured but ideally we should only do that if configs are all local otherwise we should wait for new configs from connection manager plugins.

### LINKED ISSUE #355: We can create, use, and watch a CRD in spike code

### LINKED ISSUE #356: CRDs are usable through Secretless main code

### PR DESCRIPTION
**Pull request by sgnn7**
Resolves #355 
Resolves #356
Resolves #103

[Jenkins Build](https://jenkins.conjur.net/view/cyberark/job/cyberark--secretless-broker/job/355-k8s-crds)

How to test:
- There is a full readme in [`./resource-definitions`](https://github.com/cyberark/secretless-broker/blob/bbdb2c486ebaa777aa96bbc107b6f82b6be81871/resource-definitions/README.md) that talks about the independent runners.
- If you're testing integration with the executable, secretless now can accept this type of launch:
`secretless-broker -config-mgr k8s/crd#id-of-crd-config`

Edit: All files in pkg/client are code-generated so the reviewer can skip them.

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -11,10 +11,21 @@ ENV GOOS=linux \
...
+
 # https://github.com/golang/go/issues/26610
 RUN go list -e $(go list -f '{{.Path}}' -m all 2>/dev/null)
 
 COPY . /secretless
+
+# There are checksum mismatches in varous environments with client-go package
+# so we for now manually remove it from the checksum file. Golang 1.11rc1+ seems
+# to have fixed this issue.
+RUN sed -i '/^k8s.io\/client-go\ v0.0.0-20180806134042-1f13a808da65/d' /secretless/go.sum
```
Comment (izgeri) [inline Dockerfile]: why do you have to do this twice?

File: Dockerfile.dev [hunk 2]
```diff
@@ -40,6 +45,11 @@ ENV GOOS=linux \
 
 COPY . .
 
+# There are checksum mismatches in varous environments with client-go package
+# so we for now manually remove it from the checksum file. Golang 1.11rc1+ seems
+# to have fixed this issue.
+RUN sed -i '/^k8s.io\/client-go\ v0.0.0-20180806134042-1f13a808da65/d' /secretless/go.sum
```
Comment (izgeri) [inline Dockerfile.dev]: same - why twice?

File: pkg/secretless/config/config.go [hunk 3]
```diff
@@ -165,6 +168,22 @@ func LoadFromFile(fileName string) (config Config, err error) {
 	return Load(buffer)
 }
 
+func LoadFromCRD(crdConfig crd_api_v1.Configuration) (config Config, err error) {
```
Comment (izgeri) [inline pkg/secretless/config/config.go]: how is this passing the linter? no comment

File: pkg/secretless/config/config.go [hunk 4]
```diff
@@ -165,6 +168,22 @@ func LoadFromFile(fileName string) (config Config, err error) {
...
+	var specData []byte
+	if specData, err = yaml.Marshal(crdConfig.Spec); err != nil {
+		return
+	}
+
+	log.Println(string(specData))
+	log.Println("Implement me!!")
```
Comment (izgeri) [inline pkg/secretless/config/config.go]: what's this about? :)

File: resource-definitions/README.md [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+To manually create the CRD, let's fist ensure that we have none already defined:
+```
+$ kubectl get crd
+No resources found.
+```
+
+Now lets apply our CRD file template:
+```
+$ kubectl apply -f ./secretless.yaml
```
Comment (izgeri) [inline resource-definitions/README.md]: can this be named to more clearly indicate what the file contains? maybe secretless-config-crd.yaml?

Comment (izgeri) [reply resource-definitions/README.md]: or just secretless-resourcedefinition.yaml, following the example [here](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#create-a-customresourcedefinition)

Comment (izgeri) [reply resource-definitions/README.md]: also, in that example they call `kubectl create -f resourcedefinition.yaml`, not `kubectl apply` - what's up with the difference?

File: resource-definitions/sconfig-example3.yaml [hunk 6]
```diff
@@ -0,0 +1,25 @@
+apiVersion: "secretless.io/v1"
+kind: Configuration
+metadata:
+  name: secretless-example-config2
```
Comment (izgeri) [inline resource-definitions/sconfig-example3.yaml]: 2 -> 3

File: resource-definitions/sconfig-example3.yaml [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+spec:
+  listeners:
+    - name: http_config_1_listener
```
Comment (izgeri) [inline resource-definitions/sconfig-example3.yaml]: do you want the `config_1` in the listener / handler names to be `config_2` here (and below)?

File: resource-definitions/README.md [hunk 8]
```diff
@@ -0,0 +1,127 @@
...
+
+Let us now add some configuration data:
+```
+$ kubectl apply -f ./sconfig-example.yaml
+configuration.secretless.io "secretless-example-config" created
+```
+
+We can now list and display our configuration:
+```
+$ kubectl get sconfig
```
Comment (izgeri) [inline resource-definitions/README.md]: sbconfig

File: resource-definitions/README.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+NAME                        AGE
+secretless-example-config   1m
+
+$ kubectl describe sconfig secretless-example-config
```
Comment (izgeri) [inline resource-definitions/README.md]: sbconfig

File: resource-definitions/README.md (hunk 8 above)
Comment (izgeri) [reply resource-definitions/README.md]: (there's going to be a bunch of these)

File: resource-definitions/sconfig-example3.yaml [hunk 10]
```diff
@@ -0,0 +1,25 @@
+apiVersion: "secretless.io/v1"
```
Comment (izgeri) [inline resource-definitions/sconfig-example3.yaml]: maybe this file should be called `sconfig-example2-revised` to clarify its intended use?

File: resource-definitions/README.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+2018/08/17 10:55:35 Done!
+
+$ kubectl get crd
+NAME                           AGE
+configurations.secretless.io   5s
+```
+
+# Watching CRD changes from code
+
+This code requires you to have a service account privileges but with those,
```
Comment (izgeri) [inline resource-definitions/README.md]: you should add a note about installing mercurial, and maybe even link [here](https://www.mercurial-scm.org/wiki/Download)

Comment (izgeri) [reply resource-definitions/README.md]: also add a note about sed-deleting the `client-go` go.sum line

File: resource-definitions/README.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+  generatename: ""
+  namespace: default
+  selflink: /apis/secretless.io/v1/namespaces/default/configurations/secretless-example-config
+  ...
+spec:
+  listeners:
+  ...
+  handlers:
+  ...
+```
```
Comment (izgeri) [inline resource-definitions/README.md]: are you planning to mention the two other example configs here?

File: Dockerfile.test [hunk 13]
```diff
@@ -5,11 +5,25 @@ LABEL id="secretless-test-runner"
 ENTRYPOINT [ "go", "test", "-v", "-timeout", "3m" ]
 WORKDIR /secretless
 
-RUN apk add -u curl gcc git musl-dev
+RUN apk add -u curl \
+               gcc \
+               git \
+               mercurial \
```
Comment (izgeri) [inline Dockerfile.test]: :'(

File: bin/codegen-utils/boilerplate.go.txt [hunk 14]
```diff
@@ -0,0 +1,16 @@
+/*
```
Comment (izgeri) [inline bin/codegen-utils/boilerplate.go.txt]: why is this needed?

File: Dockerfile (hunk 1 above)
Comment (izgeri) [reply Dockerfile]: we talked about this in slack - ok.

File: Dockerfile.dev (hunk 2 above)
Comment (izgeri) [reply Dockerfile.dev]: same -  ok

File: internal/pkg/plugin/manager.go [hunk 15]
```diff
@@ -35,11 +36,13 @@ func _IsDynamicLibrary(file os.FileInfo) bool {
 // Manager contains the main proxy and all connection managers, listener factories,
 // and handler factories that are available
 type Manager struct {
-	ConnectionManagers map[string]plugin_v1.ConnectionManager
-	HandlerFactories   map[string]func(plugin_v1.HandlerOptions) plugin_v1.Handler
-	ListenerFactories  map[string]func(plugin_v1.ListenerOptions) plugin_v1.Listener
-	ProviderFactories  map[string]func(plugin_v1.ProviderOptions) (plugin_v1.Provider, error)
-	Proxy              secretless.Proxy
+	reloadMutex           sync.Mutex
```
Comment (izgeri) [inline internal/pkg/plugin/manager.go]: there are a lot of things that could be reloaded - is this a `configurationReloadMutex`?

File: internal/pkg/plugin/manager.go [hunk 16]
```diff
@@ -262,9 +316,19 @@ func (manager *Manager) _RunListener(id string, options plugin_v1.ListenerOption
 }
 
 // Run is the main wait loop once we load all the plugins
-func (manager *Manager) Run(configuration config.Config) error {
+func (manager *Manager) Run(configManagerID string, filterSpec string, configuration config.Config) error {
+	// We dont want any realoads happening until we are fully running
```
Comment (izgeri) [inline internal/pkg/plugin/manager.go]: realoads -> reloads

File: bin/codegen-utils/boilerplate.go.txt (hunk 14 above)
Comment (izgeri) [reply bin/codegen-utils/boilerplate.go.txt]: we talked about this - it's so that the generated code will use our own boilerplate

File: cmd/secretless-broker/main.go [hunk 17]
```diff
@@ -3,32 +3,53 @@ package main
...
+	return
+}
+
 func main() {
 	log.Println("Secretless starting up...")
 
-	pluginDir := flag.String("p", "/usr/local/lib/secretless", "Directory containing Secretless plugins")
 	configFile := flag.String("f", "secretless.yml", "Location of the configuration file.")
+	pluginDir := flag.String("p", "/usr/local/lib/secretless", "Directory containing Secretless plugins")
+	configManagerSpec := flag.String("config-mgr", "", "Specify a configuration manager to use and filterSpec (e.g. k8s/crd#test-meta-id).")
```
Comment (izgeri) [inline cmd/secretless-broker/main.go]: please add a comment that this is optional and the default file-based config mgr will run without

File: resource-definitions/README.md [hunk 18]
```diff
@@ -0,0 +1,127 @@
+# Secretless Broker Custom Resource Definitions (CRDs)
+
+This folder contains most of the resources for dealing with CRDs in the context of
```
[the remark this answers is not in the document]
Comment (sgnn7) [reply resource-definitions/README.md]: Fixed

File: resource-definitions/README.md (hunk 11 above)
Comment (sgnn7) [reply resource-definitions/README.md]: Fixed (see pre-requisites section)

File: resource-definitions/README.md (hunk 12 above)
Comment (sgnn7) [reply resource-definitions/README.md]: Added
